lib/ui/text.dart

```diff
@@ -833,47 +833,67 @@ class TextBox {
-/// or to the right of the "H".
+/// LTR and RTL text. Consider the following string representation in code,
+/// where the lowercase letters will be displayed as LTR and the uppercase
+/// letters RTL: "helloHELLO".  When rendered, the string would be appear
+/// visually as "helloOLLEH".  An offset of 5 would be ambiguous without a
+/// corresponding TextAffinity.  Looking at the string in code, the offset
+/// represents the position just after the "o" and just before the "H".  When
+/// rendered, this offset could be either in the middle of the string to the
+/// right of the "o" or at the end of the string to the right of the "H".
 enum TextAffinity {
-  /// The position has affinity for the upstream side of the text position.
+  /// The position has affinity for the upstream side of the text position, or
+  /// in the direction of the beginning of the string in code.
   ///
-  /// For example, if the offset of the text position is at a line break, the
-  /// position represents the end of the first line.
+  /// In the example of an offset at the place where text is wrapping, upstream
+  /// indicates the end of the first line.
   ///
-  /// As another example, in the bidirectional text example above, a text
-  /// position of 6 would appear just to the right of the "o".
+  /// In the bidirectional text example above, an offset of 5 with TextAffinity
+  /// upstream would appear in the middle of the rendered text, just to the
+  /// right of the "o".
   upstream,
 
-  /// The position has affinity for the downstream side of the text position.
+  /// The position has affinity for the downstream side of the text position, or
+  /// in the direction of the end of the string in code.
   ///
-  /// For example, if the offset of the text position is at a line break, the
-  /// position represents the start of the second line.
+  /// In the example of an offset at the place where text is wrapping,
+  /// downstream indicates the beginning of the second line.
   ///
-  /// As another example, in the bidirectional text example above, a text
-  /// position of 6 would appear just to the right of the "H".
+  /// In the bidirectional text example above, an offset of 5 with TextAffinity
+  /// downstream would appear at the end of the rendered text, just to the right
+  /// of the "H".
   downstream,
 }
 
-/// A visual position in a string of text.
+/// A position in a string of text. A TextPosition can be used to locate a
+/// position in a string in code (using the [offset] property), and it can also
+/// be used to locate the same position visually in a rendered string of text
+/// (using [offset] and, when needed to resolve ambiguity, [affinity]).
 ///
-/// Must be specified along with [TextAffinity] in order to disambiguate certain
-/// situations. See the docs for TextAffinity for further details.
+/// An offset into a string in code is ambiguous with its corresponding position
```

Comment:
Here and elsewhere: you may be working too hard to distinguish a string and a string rendering. For example in this case you could just say: The location of an offset in a rendered string is ambiguous in two cases.

I made a suggestion about wording for a few of the cases.  Might be worth reconsidering the wording where "string in code" appears.